Post request

graceolivia opened the pull request:
Note: I accidentally committed a file that was too big and ran a command that seems to have affected every single commit in the history of this file. I'm sorry about all the commits.

**Story**

  Scenario: Posting echos the body
    ✔  Given I make a POST with a body to "/echo_body"              
    ✔  Then my response should have status code 200              
    ✔  And my response body should equal the body of my request  

**Changes**

- Added checkstyle hook before every commit
- Major refactor to `ListenerAndResponder`
- `HTTPRequest` and `HTTPResponse` classes, plus builder classes for each
- delete unneeded Class `RequestParser` and Test

graceolivia committed:
Create README.md

graceolivia committed:
Update README.md

graceolivia committed:
Close server socket.

graceolivia committed:
Clean up referenced files

graceolivia committed:
Current test progress

graceolivia committed:
Add mocks

graceolivia committed:
Fix Tests

WIP

WIP

Tests work

Remove unneccessary files

Remove extraneous files

Remove comments, spaces

Remove extra spaces

graceolivia committed:
Make fixes

graceolivia committed:
rename classes

graceolivia committed:
Rename test

graceolivia committed:
Cleanup Mock

graceolivia committed:
Move BUfferedReader creation logic

graceolivia committed:
Remove files

graceolivia committed:
Remove files

graceolivia committed:
Merge pull request #2 from graceolivia/gitTestBranch

Git test branch

graceolivia committed:
Add HTTP code

Add HTTP code

graceolivia committed:
Add test

graceolivia committed:
Add new logic to read in the whole HTTP

graceolivia committed:
Add non-working reader

graceolivia committed:
WIP

graceolivia committed:
Working Listener

graceolivia committed:
Ineffective attempt to fix test EchoerTest.testReadClientInputCallsPrintWriterPrintLn

graceolivia committed:
Fix mock and test

graceolivia committed:
Rename variable, remove giant import, remove override

graceolivia committed:
Reduce code duplication, remove uneccessary spaces

graceolivia committed:
Partially fixed test

graceolivia committed:
Get tests working

graceolivia committed:
Remove extra gradle file

graceolivia committed:
Merge pull request #3 from graceolivia/HTTP404

Http404

graceolivia committed:
Update README

graceolivia committed:
Add initial files

graceolivia committed:
WIP

graceolivia committed:
Add router and tests

graceolivia committed:
CHange mock

graceolivia committed:
Fix test

graceolivia committed:
Fix README

graceolivia committed:
Revert "Update README"

This reverts commit 5d4f278cf5ffd906846d24bf7a5302f99c9c9e6e.

graceolivia committed:
Minor style fixes

minor fixes'

CLRF

CLRF

graceolivia committed:
Minor style fixes

graceolivia committed:
Get enum checker of values working

graceolivia committed:
Update Echoer.java

graceolivia committed:
Fix indentation

graceolivia committed:
Add equals method

graceolivia committed:
Remove extraneous file

graceolivia committed:
Spacing

graceolivia committed:
Rename packages

graceolivia committed:
File rename

graceolivia committed:
Add loop

graceolivia committed:
Renamed Echoer and Echoable to ListenerAndResponder and ListenAndRespondable

graceolivia committed:
Merge pull request #4 from graceolivia/simpleGetFeature

Simple get feature

graceolivia committed:
Add test for 405 response to incorrect method

graceolivia committed:
Return 405 to bad request method to existing resource

graceolivia committed:
Refactor for new Router object

graceolivia committed:
WIP for fixing RequestParserTest

graceolivia committed:
Refactor StatusCodes to not include HTTP version

graceolivia committed:
Make private functions for ResponseBuilder lines

graceolivia committed:
Fix getResponse loop

graceolivia committed:
Fix typo

graceolivia committed:
Remove extra spaces and class

graceolivia committed:
Add setup for router to RouterTest

graceolivia committed:
Fix ResponseWriter method names

graceolivia committed:
Add HTTPResponse object

graceolivia committed:
Add missing test

graceolivia committed:
Rename tests

graceolivia committed:
Merge pull request #5 from graceolivia/methodNotAllowed

Method not allowed

graceolivia committed:
Add instructions for POST

graceolivia committed:
Add instructions for POST

graceolivia committed:
In Progress Commit for HTTP Request OBject Storing Headers

graceolivia committed:
In Progress Commit for HTTP Request OBject Storing Headers

graceolivia committed:
WIP on storing headers in HTTPRequest

graceolivia committed:
WIP on parsing in ListenerAndResponder

graceolivia committed:
WIP for ListenerAndResponder

graceolivia committed:
Reconfigure if statements

graceolivia committed:
Remove file

graceolivia committed:
WIP

graceolivia committed:
Merge branch 'PostRequest' of github.com:graceolivia/echo-server-2022 into PostRequest

graceolivia committed:
Create methods for RequestBuilder

graceolivia committed:
Redo ListenerAndResponer

graceolivia committed:
Proper break logic for ListenerAndResponder

graceolivia committed:
Add .hprof to gitignore

graceolivia committed:
WIP

graceolivia committed:
Begin rebuilding the ResponseBuilder

graceolivia committed:
Add setStatusLine

graceolivia committed:
WIP for HTTPResponseBuilder

graceolivia committed:
WIP for HTTPResponse

graceolivia committed:
Add checkstyle

graceolivia committed:
Add headers map printer to HTTPResponse object

graceolivia committed:
Fix test for RequestBuilder

graceolivia committed:
Add test for RequestBuilder header

graceolivia committed:
Refactor arrow-shaped code of doom WIP

graceolivia committed:
Refactor readAllLinesAndReturnHttpRequest

graceolivia committed:
Fix logic for deducing if there is a header

graceolivia committed:
Fix tests

graceolivia committed:
Fixes for tests

graceolivia committed:
Add echo body test

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
-            printServerResponse(httpResponse, clientSocket);
+            HTTPResponse httpResponse = router.getResponse(httpRequest);
+            String responseString = httpResponse.toString();
+            System.err.println(responseString.length());
```
Do you still need this statement or is it for debugging?

graceolivia replied:
Debugging! Oops, fixed

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
-    private String readAllLines(ClientReadable bufferedReader)  {
-        String line;
-        StringBuilder stringBuilder = new StringBuilder();
+    private HTTPRequest readAllLinesAndReturnHttpRequest(ClientReadable bufferedReader)  {
```
This method is very long, I'd like you to break it up into more readable fragments.

That "And" in the method name is kind of a smell that the method has more than a single responsibility.

graceolivia replied:
Broke it up! It's still coordinating reading in the body and headers and returning the HTTP Request, in more of a business logic sense.

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+            return requestBuilder.build();
+        }
+
+        if (checkIfHeadersIndicateARequestBody(headers)) {
```
This block is still extremely nested and hard to read.

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+            int bodyLength = getBodyLength(headers);
+            try {
+                while ((character = bufferedReader.read()) != null) {
+                    if (java.util.Objects.equals(bodyLength, 0)) {
```
Did `bodyLength == 0` not work here? A bit weird to be calling a static method on `java.util.Objects`.

graceolivia replied:
That works! I'm not sure why I thought it wouldn't

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+        return turnStringBuilderIntoHttpRequest(headers, body, requestBuilder);
+    }
+
+    private HTTPRequest turnStringBuilderIntoHttpRequest(StringBuilder headers, StringBuilder body, RequestBuilder requestBuilder) {
```
I like how you're moving to more descriptive variable names but I think you have a tendency to get bogged down with describing the exact implementation details rather than the abstraction. What's important to relay here? The fact that this method is using a `StringBuilder` or that this method is constructing a `HTTPRequest`?

graceolivia replied:
I guess it's that the method is constructing a `HTTPRequest` is the most important thing, and the fact that it takes in a  `StringBuilder` is obvious when you look at the parameters... Aha. So that can make the method name more terse~

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+    private boolean checkIfHeadersIndicateARequestBody(StringBuilder headers) {
+        String headersAsString = headers.toString();
+        if (headersAsString.contains("Content-Length:")) {
+            if (getBodyLength(headers) == 0) {
```
I think you can combine both conditions with `&&` in a single if clause rather than nesting them.

```java
if (headersAsString.contains("Content-Length:") && getBodyLength(headers) == 0) {
    return false
}
return true
```

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -10,8 +11,10 @@
+import java.util.Locale;
 
-import static httpServer.http.Constants.CRLF;
+import static httpServer.http.Constants.crlf;
```
Constants are typically in all upper caps; sorry that I forgot to mention before!

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+        int indexOfContentLength = indexOfContentLengthHeader + 16;
+        String trimmedHeaders = headers.substring(indexOfContentLength);
+        String lengthOfHeaders = trimmedHeaders.split(crlf)[0];
+        return Integer.parseInt(lengthOfHeaders);
```
This feels unnecessarily complicated. Can you split on `crlf`, find the element in the resulting array that contains `Content-Length` and then split that string element on `: ` to get the length?

graceolivia replied:
Fixed it... I think. It still has a loop, which is a little cumbersome, but I wasn't sure how to do something like `.forEach` with a string array

hnlee replied:
You can convert `String[]` to a `List` then use `.stream()` to convert to a `Stream` and then use `.filter()` (https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#filter-java.util.function.Predicate-) to find the right element.

graceolivia replied:
I'm having trouble with the syntax of this one, will return to this

graceolivia replied:
I ended up removing this!

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+        } else {
+            requestBuilder.buildHeaderLine(line);
+
+        }
```
I think it will be more straightforward to follow if you do:
```java
if(line.contains(": ") {
    requestBuilder.buildHeaderLine(line)
} else {
    requestBuilder.buildRequestLine(line)\
}
```

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
+        String[] headersArray = headers.toString().split(crlf);
+        for (String header: headersArray) {
+            requestBuilder = discernHeaderTypeAndAddToRequestBuilder(header, requestBuilder);
+        }
```
I feel like you can do this step with adding the headers to the request builder earlier, after line 90. Doing it earlier means that you already have that `headersArray` usable for checking for `Content-Length`. There's a lot of redundant string manipulation happening here, and I think that will cut down on it.

graceolivia replied:
Fixed this! It really did cut down on the whole process, thank you

hnlee commented on src/main/java/httpServer/http/request/HTTPRequest.java:
```diff
@@ -0,0 +1,56 @@
+        httpVersionNumber = null;
+        headers = null;
+        body = null;
+    }
```
I don't think you need to set these attributes to `null` in this constructor.

hnlee commented on src/main/java/httpServer/http/request/HTTPRequest.java:
```diff
@@ -0,0 +1,56 @@
+    public String method;
+    public String resource;
+    public String httpVersionNumber;
+    public Map<String, String> headers = new HashMap<>();
```
I wouldn't initialize the headers here.

graceolivia replied:
Would be the idea be to initialize them in a method?

hnlee replied:
I would initialize it in the public HTTPRequest() (i.e. the constructor you have that takes no parameters).

hnlee commented on src/main/java/httpServer/http/request/HTTPRequest.java:
```diff
@@ -0,0 +1,56 @@
+
+    public int getContentLength() {
+        int size = headers.size();
+        System.out.println(String.valueOf(size));
```
Do you need these above two lines or were they just for debugging?

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+    public RequestBuilder buildHeaderLine(String line) {
+        String[] requestLine = line.split(": ");
+        String removedCLDR = requestLine[1].replace(crlf, "");
+        this.headers.put(requestLine[0], removedCLDR);
```
Typo. What's `CLDR`?

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+
+    public RequestBuilder buildHeaderLine(String line) {
+        String[] requestLine = line.split(": ");
+        String removedCLDR = requestLine[1].replace(crlf, "");
```
Probably more readable to just do `String[] requestLine = line.trim().split(": ");`

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+        headers.forEach((k, v) -> System.out.println("key: " + k + ", value: " + v));
+        if (body != null) {
+            System.out.println("body: " + body);
+        }
```
Are some of these for debugging purposes?

graceolivia replied:
the ToString function itself is for debugging purposes... I don't want to dismantle it entirely just yet though, in case I need it

hnlee replied:
If so, then you'll need to write tests for it, including both test cases for with and without a body, single header, multiple headers, etc.

graceolivia replied:
I deleted it because the testing seemed like a bad use of my time and I wasn't using it anymore!

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+            System.out.println("body: " + body);
+        }
+
+        return requestLine;
```
I'm not sure if it makes sense for the string representation of `RequestBuilder` to be just the `requestLine`. Do you need this method or was it just for debugging?

graceolivia replied:
Just for debugging! I'll get it of it shortly

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+        return headers.size();
+    }
+
+    public boolean doesRequestContainBody() {
```
Might be more concise to call this just `hasBody()`

hnlee replied:
Is this a method that really belongs in `RequestBuilder` or on `HTTPRequest` itself?

graceolivia replied:
I got rid of it!

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+        return new HTTPRequest(method, resource, httpVersionNumber, headers, body);
+    }
+
+    public int length() {
```
It's not really length of `RequestBuilder`. Maybe `getNumberOfHeaders`?

hnlee replied:
Is this a method that really belongs in `RequestBuilder` or on `HTTPRequest` itself?

graceolivia replied:
Got rid of this one also

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+        return headers.containsKey("Content-Length");
+    }
+
+    public int getContentLengthInt() {
```
Is this a method that really belongs in `RequestBuilder` or on `HTTPRequest` itself?

graceolivia replied:
I need it because I am using the `RequestBuilder` to determine the amount of characters I should read in in `ListenerAndResponder.readBody`, as far as I can tell

hnlee commented on src/main/java/httpServer/http/request/RequestBuilder.java:
```diff
@@ -0,0 +1,81 @@
+        System.out.println(expectedBodyLength);
+        System.out.println(currentBodyLength);
+        System.out.println(body);
+        return (expectedBodyLength == currentBodyLength);
```
This looks more like a test...

graceolivia replied:
Deleted!

hnlee commented on src/main/java/httpServer/http/response/HTTPResponse.java:
```diff
@@ -0,0 +1,56 @@
+        }
+        return body;
+
+    }
```
whitespace

hnlee commented on src/main/java/httpServer/http/response/HTTPResponse.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private StringBuilder appendIfNotNull(StringBuilder stringBuilder, String lineOrHeader) {
+        if (!(lineOrHeader == null)) {
```
You can just do `lineOrHeader != null`

hnlee commented on src/main/java/httpServer/http/response/HTTPResponse.java:
```diff
@@ -0,0 +1,56 @@
+
+    }
+
+    private StringBuilder appendHeaders(StringBuilder stringBuilder, Map<String, String> headerMap) {
```
Do you need this?

graceolivia replied:
No, deleted

hnlee commented on src/main/java/httpServer/http/response/HTTPResponseBuilder.java:
```diff
@@ -0,0 +1,67 @@
+
+    public HTTPResponseBuilder makeAllowHeader(List methods) {
+        StringBuilder allowedMethods = new StringBuilder();
+        for (int i = 0; i < methods.size(); i++) {
```
You can use `String.join` to just create the string from the `methods` array, without using a `for` loop.

graceolivia replied:
Ended up using a stream in a similar way to make the whole thing shorter!

hnlee commented on src/main/java/httpServer/http/response/HTTPResponseBuilder.java:
```diff
@@ -0,0 +1,67 @@
+            this.headers.put("Content-Length", "0");
+        } else {
+            int bodyLength = body.length();
+            this.headers.put("Content-Length", String.valueOf(bodyLength));
```
You can just do `this.headers.put("Content-Length", String.valueOf(body.length));`

hnlee commented on src/main/java/httpServer/http/response/HTTPResponseBuilder.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private String setContentLengthHeader(int length) {
+        return ("Content-Length: " + String.valueOf(length));
```
Do you need this?

graceolivia replied:
Nope! It's deleted

hnlee commented on src/main/java/httpServer/routes/Router.java:
```diff
@@ -12,29 +13,60 @@
+        HTTPMethods[] allowedRoutes = routes.get(httpRequest.resource);
+        List allowedRoutesAsList = Arrays.asList(allowedRoutes);
+
+        for (int i = 0; i < allowedRoutesAsList.size(); i++) {
```
Try to find a way to do this without using a `for` loop.

hnlee commented on src/test/java/httpServer/ListenerAndResponderTest.java:
```diff
@@ -22,18 +22,18 @@
-                "User-Agent: jUnit" + Constants.CRLF +
-                "Host: localhost:5000" + Constants.CRLF +
-                "GET / HTTP/1.1" + Constants.CRLF + Constants.CRLF;
+        String input = "Accept: */*" + Constants.crlf +
```
I think if you do `import static httpServer.http.Constants.crlf` you don't have to qualify the constant each time.

hnlee commented on src/test/java/httpServer/ListenerAndResponderTest.java:
```diff
@@ -55,7 +55,7 @@ void testThatStartServerReturnsASocket() throws IOException {
     @Test
     void testReadClientInputCallsPrintWriterPrintLn() throws IOException {
-        String expectedResult = "HTTP/1.1 " + StatusCodes.PAGE_NOT_FOUND.httpResponse + Constants.CRLF + "Content-Length: 0" + Constants.CRLF;
+        String expectedResult = "HTTP/1.1 " + StatusCodes.PAGE_NOT_FOUND.httpResponse + Constants.crlf + "Content-Length: 0" + Constants.crlf;
```
Might want to fix this test, since you're no longer using `println`

hnlee commented on src/test/java/httpServer/MockBufferedReaderWrapper.java:
```diff
@@ -17,7 +17,7 @@ public MockBufferedReaderWrapper(String input) {
-    public String readLine() throws IOException {
+    public String read() throws IOException {
         readLineWasCalled = true;
         if ( inputStack.size() > 0) {
```
whitespace

hnlee commented on src/test/java/httpServer/MockPrintWriterWrapper.java:
```diff
@@ -7,14 +7,19 @@
 
     @Override
-    public void println(String message) throws IOException {
+    public void print(String message) throws IOException {
```
`printlnWasCalledWith` should become `printWasCalledWith`

hnlee commented on src/test/java/httpServer/RequestBuilderTest.java:
```diff
@@ -0,0 +1,55 @@
+
+        HTTPRequest httpParsed = requestBuilder.build();
+        assertEquals(httpRequestExpected.headers.get("Host"), httpParsed.headers.get("Host"));
+    }
```
Test for multiple headers? Test for body?

graceolivia replied:
Added these!

hnlee reviewed (changes requested):
You have a lot more logic now in `ListenerAndResponder` which is going untested. I wonder if some of the struggle you had with this feature might have been avoided if you had written tests for the body length logic.

Not a blocker for this PR but `ListenerAndResponder` may need to get split into smaller classes.

You don't have any tests for `HTTPResponseBuilder`.

graceolivia committed:
Minor fixes to ListenerAndResponder

graceolivia committed:
Break readHeaderAndBodyAndReturnHttpRequest into modules

graceolivia committed:
Test import

graceolivia committed:
Test import

graceolivia committed:
Uppercase CRLF

graceolivia committed:
Refactor GetBodyLength

graceolivia committed:
Remove unneeded HTTPResponse methods

graceolivia committed:
Add HEAD

graceolivia committed:
Fix typo

graceolivia committed:
delete some unused methods

graceolivia committed:
Add body and multiple header tests

graceolivia committed:
WIP for stream

graceolivia committed:
Remove unneccessary assignments

graceolivia committed:
Remove unneccessary methods from HTTPResponseBuilder

graceolivia committed:
Fix setContentLengthHeader to be more terse

graceolivia committed:
Refactor makeAllowHeader

graceolivia committed:
WIP for isRouteInExistingResource loop

graceolivia committed:
Fix tests, refactor ListenerAndResponder

remove print statements

Remove extra method

Remove loop from isRouteInExistingResource

Add tests for Response Builder

Add the test

Remove comment

graceolivia committed:
Add test for ListenerAndResponder

graceolivia committed:
Remove stray print statement

graceolivia committed:
Add tests for ListenerAndResponder

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +40,102 @@ public Socket keepListening() throws IOException {
 
     public void readClientInput(Socket clientSocket) throws IOException {
-
+        System.err.println("Starting to read");
```
Don't forget to remove debugging statements!

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +40,102 @@ public Socket keepListening() throws IOException {
-        String httpRequest;
-        httpRequest = readAllLines(bufferedReader);
+        HTTPRequest httpRequest;
+        httpRequest = readHeaderAndBodyAndReturnHttpRequest(bufferedReader);
```
I think I pointed out in my last review that using `and` in your method or class names is probably a violation of single responsibility principle. In this case, it doesn't necessarily seem like a violation but I would recommend renaming this method anyway. What is its responsibility?

hnlee replied:
As an additional note, you can just combine lines 45-46 in one line:
```java
HTTPRequest httpRequest = readHeaderAndBodyAndReturnHttpRequest(bufferedReader);
```

graceolivia replied:
It turns the String that get sent into a HTTPRequest object, so I renamed accordingly

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +40,102 @@ public Socket keepListening() throws IOException {
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
```
On the one hand I'm glad you internalized the "more verbose variable names are clearer" but in this case, I don't think it's actually adding to clarity. Maybe we can just call `checkIfReachedEndOfHeaders` to `atEndOfHeaders()`

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +40,102 @@ public Socket keepListening() throws IOException {
         }
+        String[] headersArray = headers.toString().split(CRLF);
+        for (String header: headersArray) {
+            requestBuilder = discernHeaderTypeAndAddToRequestBuilder(header, requestBuilder);
```
This is a good example of two responsibilities being reflected in the name.

You probably want a `isRequestLine` method.

You could then probably do something like:
```java
requestBuilder = isRequestLine? requestBuilder.buildRequestLine(header) : requestBuilder.buildHeaderLine(header);
```

I think it's also misleading to call it `headers`, `headersArray` and `header` when you're not sure if it's a header. Maybe we could call it `metadata`, `metadataLines` and `metadataLine`?

hnlee commented on src/main/java/httpServer/ListenerAndResponder.java:
```diff
@@ -40,46 +40,102 @@ public Socket keepListening() throws IOException {
+
+    private boolean checkIfReachedEndOfHeaders(StringBuilder requestScannedInSoFar) {
+        return requestScannedInSoFar.toString().contains(CRLF + CRLF);
+    }
```
There's overall a _lot_ of private methods relating to the reading in of the request that feels like maybe it should be in its own class. I won't make it a condition for merge of this PR, but I will ask you to do some extraction for your next PR.

hnlee commented on src/main/java/httpServer/http/response/HTTPResponse.java:
```diff
@@ -0,0 +1,39 @@
+public class HTTPResponse {
+
+    public String statusLine;
+    public Map<String, String> headers = new HashMap<>();
```
Let's match `HTTPRequest` and not initialize an empty `HashMap` here.

hnlee commented on src/main/java/httpServer/routes/Router.java:
```diff
@@ -1,40 +1,68 @@
+        return routes.containsKey(httpRequest.resource);
+    }
+
+    private boolean isRouteInExistingResource(HTTPRequest httpRequest) {
```
Are you still using this method?

graceolivia replied:
Yes I am, in `setStatusCodeLine`

hnlee commented on src/test/java/httpServer/RequestBuilderTest.java:
```diff
@@ -0,0 +1,72 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class RequestBuilderTest {
```
It's a little confusing because the class is called `HTTPRequestBuilder` but the test is called `RequestBuilderTest`.

hnlee commented on src/test/java/httpServer/RequestBuilderTest.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @Test
+    void testRequestBuilderCorrectlyParsesStatusLine() throws IOException {
```
It's request line for requests and status line for responses.

hnlee reviewed (changes requested):
I know you've struggled a lot with this PR, and it may be dispiriting to be asked to make more changes, but I think overall, it's looking very good! `ListenerAndResponder` does need considerable cleanup, but outside of that particular class, the code is actually very clean and clear and easy to follow.

graceolivia committed:
Make changes from PR

graceolivia committed:
Remove unused method

hnlee commented on src/main/java/httpServer/HttpServer.java:
```diff
@@ -44,6 +44,7 @@ public static void closeSockets(Socket clientSocket, ServerSocketInterface serve
         routes.put("/simple_get", new HTTPMethods[] {HTTPMethods.GET, HTTPMethods.OPTIONS, HTTPMethods.HEAD});
         routes.put("/head_request", new HTTPMethods[] {HTTPMethods.HEAD, HTTPMethods.OPTIONS});
         routes.put("/echo_body", new HTTPMethods[] {HTTPMethods.HEAD, HTTPMethods.OPTIONS, HTTPMethods.POST});
+        routes.put("/json_response", new HTTPMethods[] {HTTPMethods.GET, HTTPMethods.OPTIONS});
```
I think this belongs to your other branch?

hnlee commented on src/main/java/httpServer/http/response/HTTPResponseBuilder.java:
```diff
@@ -39,8 +39,12 @@ public HTTPResponseBuilder setContentLengthHeader() {
+            this.headers.put("Content-Type", "application/json");
+        } else {
+            this.headers.put("Content-Type", "text/plain");
+        }
```
Looks like this belongs on a different branch too.

hnlee commented on src/main/java/httpServer/routes/Router.java:
```diff
@@ -23,7 +23,8 @@ public Router(Map routes, HTTPResponseBuilder responseBuilder) {
         responseBuilder = setStatusCodeLine(httpRequest, responseBuilder);
-        responseBuilder = responseBuilder.setBody(httpRequest.body);
+        responseBuilder = setBody(httpRequest, responseBuilder);
+        //responseBuilder.setBody("{}");
```
This might need deleting?

hnlee commented on src/main/java/httpServer/routes/Router.java:
```diff
@@ -47,7 +48,20 @@ private HTTPResponseBuilder setHeaders(HTTPRequest httpRequest, HTTPResponseBuil
+        System.err.println("SETBODY");
+        if (httpRequest.resource.equals("/json_response")) {
+            System.err.println("JASON");
+            responseBuilder = responseBuilder.setBody("{\"people\": [{\"craft\": \"ISS\", \"name\": \"Mark Vande Hei\"}, {\"craft\": \"ISS\", \"name\": \"Pyotr Dubrov\"}, {\"craft\": \"ISS\", \"name\": \"Anton Shkaplerov\"}, {\"craft\": \"Shenzhou 13\", \"name\": \"Zhai Zhigang\"}, {\"craft\": \"Shenzhou 13\", \"name\": \"Wang Yaping\"}, {\"craft\": \"Shenzhou 13\", \"name\": \"Ye Guangfu\"}, {\" ...
```
This too...

hnlee reviewed (approved):
You may want to remove the changes relating to your JSON response feature from this PR! No need to ask for re-approval; just merge after you clean up.

graceolivia committed:
fix things from spike
